File: crates/prolly/bindings/wasm/examples/resolver.ts
import assert from "node:assert/strict";
import { existsSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { loadProllyWasm } from "../src/index.ts";

const enc = new TextEncoder();
const dec = new TextDecoder();
const here = dirname(fileURLToPath(import.meta.url));
const pkgJs = join(here, "../pkg/prolly_wasm.js");
const pkgWasm = join(here, "../pkg/prolly_wasm_bg.wasm");

const bytes = (value: string): Uint8Array => enc.encode(value);
const text = (value: Uint8Array | null | undefined): string | null =>
  value == null ? null : dec.decode(value);
const upsert = (key: string, value: string | Uint8Array) => ({
  kind: "upsert",
  key: bytes(key),
  value: typeof value === "string" ? bytes(value) : value,
});
const del = (key: string) => ({ kind: "delete", key: bytes(key) });

async function loadWasm() {
  if (!existsSync(pkgJs) || !existsSync(pkgWasm)) {
    throw new Error("WASM package is not built. Run npm --prefix crates/prolly/bindings/wasm run build:wasm first.");
  }
  return loadProllyWasm("../pkg/prolly_wasm.js", readFileSync(pkgWasm));
}

async function resolver(): Promise<void> {
  const wasm = await loadWasm();
  const engine = wasm.WasmProllyEngine.memory();
  const base = engine.put(engine.create(), bytes("settings/theme"), bytes("light"));
  const leftDelete = engine.delete(base, bytes("settings/theme"));
  const rightUpdate = engine.put(base, bytes("settings/theme"), bytes("dark"));

  const updateWins = engine.merge(base, leftDelete, rightUpdate, "update_wins");
  const deleteWins = engine.merge(base, leftDelete, rightUpdate, "delete_wins");

  assert.equal(text(engine.get(updateWins, bytes("settings/theme"))), "dark");
  assert.equal(engine.get(deleteWins, bytes("settings/theme")), null);

  console.log("resolver: demonstrated update-wins and delete-wins policies");
}

await resolver();

File: crates/prolly/examples/resolver.rs
use prolly::{resolver, Config, MemStore, Prolly, Resolution, Resolver};

fn main() -> Result<(), prolly::Error> {
    let prolly = Prolly::new(MemStore::new(), Config::default());

    let base = prolly.create();
    let base = prolly.put(&base, b"setting:theme".to_vec(), b"system".to_vec())?;

    let left = prolly.delete(&base, b"setting:theme")?;
    let right = prolly.put(&base, b"setting:theme".to_vec(), b"dark".to_vec())?;

    let update_wins = prolly.merge(&base, &left, &right, Some(Box::new(resolver::update_wins)))?;
    assert_eq!(
        prolly.get(&update_wins, b"setting:theme")?,
        Some(b"dark".to_vec())
    );

    let delete_wins = prolly.merge(&base, &left, &right, Some(Box::new(resolver::delete_wins)))?;
    assert_eq!(prolly.get(&delete_wins, b"setting:theme")?, None);

    let leave_unresolved: Resolver = Box::new(|conflict| {
        if conflict.key.starts_with(b"setting:") {
            Resolution::unresolved()
        } else {
            resolver::prefer_right(conflict)
        }
    });

    assert!(prolly
        .merge(&base, &left, &right, Some(leave_unresolved))
        .is_err());

    println!("delete/update conflict can resolve to value, delete, or error");
    Ok(())
}
